learn/functions/recursion.md

```diff
@@ -0,0 +1,124 @@
+
+## Common mistakes and pitfalls in Python Recursion
+
+Python recursive functions can be a powerful tool in solving complex problems, but they can also be prone to common mistakes and pitfalls. Here are some common mistakes to avoid when using python recursion:
+
+1. **Infinite loops**: It is easy to create infinite loops in recursive functions if you don't have a base case that ultimately breaks the recursion. Be sure to define a base case that will stop the recursion.
+
+2. **Stack overflow**: Recursion can create a lot of memory overhead, which can lead to stack overflow errors. Be careful when using recursion with large datasets.
+
+## Enhancing Python recursive functions for greater efficiency and performance
+
+Recursive enhancing refers to the process of optimizing a Python recursive function for better efficiency and performance. This involves identifying areas that can be fine-tuned, such as reducing space complexity or using memoization to reduce the number of recursive calls.
+
+Below are two examples of how to enhance Python recursive functions for greater efficiency:
+
+### Memoization
+
+Memoization is the process of storing previously computed results to avoid repeated calculation. This can significantly reduce the runtime of a recursive function. 
+
+```python
+def fibonacci(n, memo={}):
+    if n <= 1:
+        return n
+    elif n in memo:
+        return memo[n]
+    else:
+        memo[n] = fibonacci(n-1, memo) + fibonacci(n-2, memo)
+        return memo[n]
+```
+
+In the above code, the `memo` dictionary is used to store previously computed Fibonacci numbers. When the function is called with a previously calculated `n`, the `memo` value is returned instead of the function making another recursive call.
+
+### Tail Recursion Optimization
+
+Tail recursion optimization is a way to optimize recursive functions so that they use less space on the call stack. 
+
+```python
+def sum_n(n, acc=0):
+    if n == 0:
+        return acc
+    else:
+        return sum_n(n-1, acc+n)
+
+# Example usage
+print(sum_n(5)) # Outputs: 15
+```
+
+In the above code, `sum_n()` is a **recursive function** that calculates the sum of all numbers from `1` to `n`. The acc argument is an **accumulator** that stores the intermediate results of the computation.
+
+In each recursive call, the function adds the current value of `n` to the accumulator and passes the result to the next recursive call, without keeping the previous call's stack frame in memory. This way, the function uses a constant amount of memory on the call stack, and avoids the risk of stack overflow for large values of `n`.
+
+Note that tail recursion optimization can only be applied to recursive functions that have a tail call, i.e., a recursive call that occurs at the end of the function's execution.
```

Comment:
For note text we can use `> ` markup - https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-text